drafts/0007-RFC-BAG-OPERATORS.md

```diff
@@ -0,0 +1,564 @@
+  - If A or B is << >>, then << >>
+  - Else << x >> with multiplicity m, for all x in A with x in B and m = min(M(x, A), M(x, B))
+
+EXCEPT_ALL(A, B):
+  - If A is << >>, then << >>
+  - If B is << >>, then A
+  - Else << x >> with multiplicity m, for all x in A with m = max(M(x, A) - M(x, B), 0)
+
+UNION(A, B) = DISTINCT(UNION_ALL(A, B))
+
+INTERSECT(A, B) = DISTINCT(INTERSECT_ALL(A, B))
+
+EXCEPT(A, B) = DISTINCT(EXCEPT_ALL(A, B))
+```
+
+The following examples show how these operators can be used to construct new heterogenous bags. All examples will use permissive typing mode as conventional typing mode is equivalent to the SQL bag operators.
+
+### Example Environment
+
+TODO: I would like some feedback on this example data. In the real world, we don't always control the shape of our data; but a reader may look at this and think "why isn't this flat if employees is homogenous?". As a database schema, this would be an "Employees" table and a "Department" table with a separate relation table. I would like some feedback on example data shape that would avoid this initial reaction.
+
+```
+{ 
+  'hr': { 
+    'employees': <<
+      { 'id': 1, 'name': 'Bob Smith',   'title': 'Consultant' }, 
+      { 'id': 2, 'name': 'Susan Smith', 'title': 'Manager' },
+      { 'id': 3, 'name': 'Jane Smith',  'title': 'DEI'}
+    >>
+  },
+  'accounting': { 
+    'employees': <<
+      { 'id': 1, 'name': 'Quincy Jones',   'title': 'Director' }, 
+      { 'id': 2, 'name': 'James Early Jones', 'title': 'Manager' },
+      { 'id': 3, 'name': 'Indiana Jones',  'title': 'Intern'}
+    >>
+  },
+  'engineering': { 
+    'employees': <<
+      { 'id': 1, 'name': 'Paul McCharmly',   'title': 'Manager' }, 
+      { 'id': 2, 'name': 'George Parasol', 'title': 'SDE' },
+      { 'id': 3, 'name': 'Ringo Stone',  'title': 'Principal SDE'},
+      { 'id': 4, 'name': 'Eric Lennon',  'title': 'Intern'}
+    >>
+  }
+} 
+```
+
+### Union Examples
+
+**Union 1** — No manager meeting invite list
+```
+PartiQL> SELECT id, name FROM (
+   |   SELECT * FROM hr.employees
+   |   UNION
+   |   SELECT * FROM accounting.employees
+   |   UNION
+   |   SELECT * FROM engineering.employees
+   | )
+   | WHERE title != 'Manager'
+   | ORDER BY title ASC
+   |
+==='
+[
+  {
+    'id': 1,
+    'name': 'Bob Smith'
+  },
+  {
+    'id': 3,
+    'name': 'Jane Smith'
+  },
+  {
+    'id': 1,
+    'name': 'Quincy Jones'
+  },
+  {
+    'id': 3,
+    'name': 'Indiana Jones'
+  },
+  {
+    'id': 4,
+    'name': 'Eric Lennon'
+  },
+  {
+    'id': 3,
+    'name': 'Ringo Stone'
+  },
+  {
+    'id': 2,
+    'name': 'George Parasol'
+  }
+]
+---
+OK!
+```
+
+**Union 2** — Union where right operands are expressions rather than SFW blocks. Note that this is stated as possible in the SQL++ text, but the grammar does not support this. This is the result of an initial implementation because our parser actually parsers UNION operators as expressions rather than SFW blocks as described by the grammar.
```

Comment:
> but the grammar does not support this
What grammar?
> Note that this is stated as possible in the SQL++ text ... our parser actually parsers UNION operators as expressions rather than SFW blocks as described by the grammar.
I'm not sure what the problem is?  Is it even a problem?

Reply:
The grammar specified in the PartiQL spec states that the left and right operand of UNION|INTERSECT|EXCEPT is a */<sfw_query/>*, but the parser actually parsers these as expressions. I'm not sure it's a problem -- just not exactly an explicit feature that could lead to confusion

```
PartiQL> SELECT * FROM << 1, 2, 3 >> UNION << 4, 5, 6 >>
   |
==='
<<
  {
    '_1': 1
  },
  {
    '_1': 2
  },
  {
    '_1': 3
  },
  4,
  5,
  6
>>
---
```

Compared to
```
PartiQL> SELECT * FROM << 1, 2, 3 >>
   | UNION ALL
   | SELECT * FROM << 4, 5, 6 >>
   |
==='
<<
  {
    '_1': 1
  },
  {
    '_1': 2
  },
  {
    '_1': 3
  },
  {
    '_1': 4
  },
  {
    '_1': 5
  },
  {
    '_1': 6
  }
>>
---
OK!
```

Reply:
Any comment on this? I'm leaning towards changing the parser to be consistent with the specification.